### PR TITLE
Update documentation for Global and clean up usage of globalStatus

### PR DESCRIPTION
**Pull request by btracey**

### CODE REVIEW

**Comment by btracey** [comment]
This changes the wording somewhat from the Local documentation, I can rectify the differences when we like the Global wording.

File: global.go [hunk 1]
```diff
@@ -183,52 +210,109 @@ func globalWorker(task int, m GlobalMethod, g *globalStatus, loc *Location, x []
...
+}
+
+func (g *globalStatus) finishIteration(status Status, err error, loc *Location, op Operation) (Status, error) {
+	g.mux.Lock()
+	defer g.mux.Unlock()
+	return finishIteration(status, err, g.stats, g.settings, g.statuser, g.startTime, loc, op)
+}
+
+// globalOperation executes the requested operation at the given location.
+// Modifications to this function must keep in mind that it can be called
```
Comment (sbinet) [inline global.go]: `"Modifications to this function must keep in mind that it..."`

this reads a bit funny.

perhaps:
`Modifications to this function must be applied while keeping in mind that globalOperation can be...`

Comment (btracey) [inline global.go]: Done.

File: global.go [hunk 2]
```diff
@@ -10,17 +10,54 @@ import (
...
+	// IterateGlobal retrieves information from the location associated with
+	// the given task ID, and returns the next operation to perform with that
+	// Location. IterateGlobal may assume that the same pointer is associated
+	// with the same task.
 	IterateGlobal(task int, loc *Location) (Operation, error)
-	Needser
-	// Done communicates to the optimization method that the optimization has
-	// concluded to allow for shutdown.
+
+	// Done communicates that the optimization has concluded to allow for shutdown.
```
Comment (sbinet) [inline global.go]: perhaps semantically relate it to the `sync.WaitGroup.Done()` method? (or explicitly unrelate it to that)

I was a bit thrown off tracks that it didn't return anything (not even `struct{}`) so it couldn't really be used like in:
```go
for {
    select {
    case <-gm.Done(): // do clean-up
    }
}
```
and then I remembered that it looked a bit like `sync.WaitGroup.Done()`.

File: global.go [hunk 3]
```diff
@@ -148,30 +191,14 @@ func minimizeGlobal(p *Problem, method GlobalMethod, settings *Settings, stats *
...
-	p         *Problem
-	startTime time.Time
-	optLoc    *Location
-	settings  *Settings
-	method    GlobalMethod
-	statuser  Statuser
-	err       error
-}
-
+// globalWorker runs the optimization stps for a single (concurrently-executing)
```
Comment (vladimir-ch) [inline global.go]: steps

File: global.go [hunk 4]
```diff
@@ -183,52 +210,109 @@ func globalWorker(task int, m GlobalMethod, g *globalStatus, loc *Location, x []
 	}
 }
 
-// globalOperation updates handles the status received by an individual worker.
-// It uses a mutex to protect updates where necessary.
-func (g *globalStatus) globalOperation(op Operation, loc *Location, x []float64) Status {
-	// Do a quick check to see if one of the other workers converged in the meantime.
+// globalStatus is a data struct that coordinates information between concurrently
```
Comment (vladimir-ch) [inline global.go]: `globalStatus coordinates access to information shared between ...`?

File: global.go [hunk 5]
```diff
@@ -183,52 +210,109 @@ func globalWorker(task int, m GlobalMethod, g *globalStatus, loc *Location, x []
...
+// updateStats updates the evaluation statistics for the given operation.
+func (g *globalStatus) updateStats(op Operation) {
+	g.mux.Lock()
+	defer g.mux.Unlock()
+	updateEvaluationStats(g.stats, op)
+}
+
+// updateStatus updates the status and error fields of g. This update only happens
+// if status == NotTerminated, so that the first different status is the one
+// maintained
```
Comment (vladimir-ch) [inline global.go]: Dot at the end.

File: global.go [hunk 6]
```diff
@@ -183,52 +210,109 @@ func globalWorker(task int, m GlobalMethod, g *globalStatus, loc *Location, x []
...
+func (g *globalStatus) finishIteration(status Status, err error, loc *Location, op Operation) (Status, error) {
+	g.mux.Lock()
+	defer g.mux.Unlock()
+	return finishIteration(status, err, g.stats, g.settings, g.statuser, g.startTime, loc, op)
+}
+
+// globalOperation executes the requested operation at the given location.
+// When modifying this function, keep in mind that it can be called concurrently.
+// Uses of the internal fields should be through the methods of globalStatus and
+// protected by a mux where appropriate.
```
Comment (vladimir-ch) [inline global.go]: mutex

File: global.go [hunk 7]
```diff
@@ -183,52 +210,109 @@ func globalWorker(task int, m GlobalMethod, g *globalStatus, loc *Location, x []
...
+	// Do a quick check to see if one of the other workers converged in the meantime.
+	status := g.getStatus()
 	if status != NotTerminated {
 		return status
 	}
+	var err error
 	switch op {
 	case NoOperation:
 	case InitIteration:
 		panic("optimize: Method returned InitIteration")
```
Comment (vladimir-ch) [inline global.go]: GlobalMethod

File: global.go [hunk 8]
```diff
@@ -183,52 +210,109 @@ func globalWorker(task int, m GlobalMethod, g *globalStatus, loc *Location, x []
...
 	case PostIteration:
 		panic("optimize: Method returned PostIteration")
```
Comment (vladimir-ch) [inline global.go]: GlobalMethod

**Comment by btracey** [comment]
Done on all requests.

**Comment by btracey** [comment]
Implemented by https://github.com/gonum/gonum/pull/265

**Pull request closed without merging**